### PR TITLE
feat: Ability to retry trigger

### LINKED ISSUE #1089: More logging for event processing
**Is your feature request related to a problem? Please describe.**
e.g. events got dropped 

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by whynowy**
Signed-off-by: Derek Wang <whynowy@gmail.com>

The PR:

1. Option to specify `retryStrategy` for a trigger;
2. Refined `common.Backoff` struct, make it possible to use strings such as `2s`, `1m` as it's `Duration`, meanwhile make it backward compatible to support nanoseconds;
3. Added more logs to both EventSource and Sensor, this is to close #1089.

Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-events/blob/master/USERS.md).

### CODE REVIEW

File: pkg/apis/common/common.go [hunk 1]
```diff
@@ -103,24 +103,28 @@ type TLSConfig struct {
...
 	// Deprecated: use ClientKeySecret instead
-	DeprecatedClientKeyPath string `json:"clientKeyPath" protobuf:"bytes,6,opt,name=clientKeyPath"`
+	DeprecatedClientKeyPath string `json:"clientKeyPath,omitempty" protobuf:"bytes,6,opt,name=clientKeyPath"`
 }
 
 // Backoff for an operation
 type Backoff struct {
-	// Duration is the duration in nanoseconds
-	Duration time.Duration `json:"duration" protobuf:"varint,1,opt,name=duration,casttype=time.Duration"`
+	// The initial duration such as "1s", "3m"
```
Comment (whynowy) [inline pkg/apis/common/common.go:117]: Update the spec to make it possible to only specify `steps`.

File: sensors/artifacts/store.go [hunk 2]
```diff
@@ -34,11 +35,14 @@ type ArtifactReader interface {
 
 // FetchArtifact from the location, decode it using explicit types, and unstructure it
 func FetchArtifact(reader ArtifactReader) (*unstructured.Unstructured, error) {
-	var err error
 	var obj []byte
-	obj, err = reader.Read()
-	if err != nil {
-		return nil, err
+
+	if err := common.Connect(&common.DefaultRetry, func() error {
```
Comment (whynowy) [inline sensors/artifacts/store.go:40]: Add retry to fetch external resources such as getting workflow spec from S3.

File: sensors/listener.go [hunk 3]
```diff
@@ -284,25 +286,29 @@ func (sensorCtx *SensorContext) triggerOne(ctx context.Context, sensor *v1alpha1
...
-		sensorCtx.metrics.ActionFailed(sensor.Name, trigger.Template.Name)
 		return err
 	}
 
 	log.Debugw("executing the trigger resource", "triggerName", trigger.Template.Name)
-	newObj, err := triggerImpl.Execute(ctx, eventsMapping, updatedObj)
-	if err != nil {
-		sensorCtx.metrics.ActionFailed(sensor.Name, trigger.Template.Name)
-		return err
+	retryStrategy := trigger.RetryStrategy
```
Comment (whynowy) [inline sensors/listener.go:293]: Trigger execution retry.

File: pkg/apis/common/common.go [hunk 4]
```diff
@@ -103,24 +101,28 @@ type TLSConfig struct {
...
 	// Deprecated: use ClientKeySecret instead
-	DeprecatedClientKeyPath string `json:"clientKeyPath" protobuf:"bytes,6,opt,name=clientKeyPath"`
+	DeprecatedClientKeyPath string `json:"clientKeyPath,omitempty" protobuf:"bytes,6,opt,name=clientKeyPath"`
 }
 
 // Backoff for an operation
 type Backoff struct {
-	// Duration is the duration in nanoseconds
-	Duration time.Duration `json:"duration" protobuf:"varint,1,opt,name=duration,casttype=time.Duration"`
+	// The initial duration in nanoseconds or strings like "1s", "3m"
```
Comment (whynowy) [inline pkg/apis/common/common.go:115]: The goal of changing `Backoff` properties:

1. Make it easy to use, user can use strings like `2s`, `1m` for `Duration`;
2. Make it possible to only specify a property such as `steps`, and use default ones for the rest.

File: common/retry.go [hunk 5]
```diff
@@ -42,38 +48,56 @@ func IsRetryableKubeAPIError(err error) bool {
...
+	if jitter == nil {
+		jitter = &defaultJitter
 	}
-	result.Duration = backoff.Duration
-	result.Factor, _ = backoff.Factor.Float64()
-	if backoff.Jitter != nil {
-		result.Jitter, _ = backoff.Jitter.Float64()
+	j, err := jitter.Float64()
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid jitter")
```
Comment (alexec) [inline common/retry.go:77]: minor - but I'm not a fan, I prefer `fmt.Errorf("invalid jitter: %w", err)`

File: pkg/apis/common/common.go [hunk 6]
```diff
@@ -103,24 +101,28 @@ type TLSConfig struct {
...
+	DeprecatedClientKeyPath string `json:"clientKeyPath,omitempty" protobuf:"bytes,6,opt,name=clientKeyPath"`
 }
 
 // Backoff for an operation
 type Backoff struct {
-	// Duration is the duration in nanoseconds
-	Duration time.Duration `json:"duration" protobuf:"varint,1,opt,name=duration,casttype=time.Duration"`
+	// The initial duration in nanoseconds or string like "1s", "3m"
+	// +optional
+	Duration *Int64OrString `json:"duration,omitempty" protobuf:"varint,1,opt,name=duration"`
```
[the remark this answers is not in the document]
Comment (alexec) [reply pkg/apis/common/common.go:117]: and get ready to have many bugs...

**Review by alexec — APPROVED** [review]

**Comment by alexec** [comment]
do you need int64? if not, use the intorstr

File: pkg/apis/common/common.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (whynowy) [reply pkg/apis/common/common.go:117]: :)

**Review by whynowy — COMMENTED** [review]
> do you need int64? if not, use the intorstr

The small 10s in nanoseconds is larger than `MaxInt32`, so I have to.

**Pull request merged**